File: gfapy/line/edge/link/complement.py
import gfapy

class Complement:

  def complement(self):
    """Creates the equivalent link with from and to inverted.

    The CIGAR operations (order and type) are inverted as well.
    Tags are left unchanged.

    Note:
        The path references are not copied to the complement link.

    Note:
        This method shall be overridden if custom tags are defined, which have a
        complementation operation which determines their value in the
        equivalent complement link.

    Returns:
        gfapy.line.edge.Link: The inverted link.
    """
    l = self.clone()
    l.from_segment = self.to_segment
    l.from_orient = gfapy.invert(self.to_orient)
    l.to_segment = self.from_segment
    l.to_orient = gfapy.invert(self.from_orient)
    l.overlap = self.overlap.complement()
    return l

  def make_complement(self):
    """Complements the link inplace.

    The tags are left unchanged.

    Note:
        The path references are not complemented by this method; therefore
        the method shall be used before the link is embedded in a graph.

    Note:
        This method shall be overridden if custom tags are defined, which have a
        complementation operation which determines their value in the
        complement link.

    Returns:
        gfapy.line.edge.Link: self
    """
    tmp = self.from_segment
    self.from_segment = self.to_segment
    self.to_segment = tmp
    tmp = self.from_orient
    self.from_orient = gfapy.invert(self.to_orient)
    self.to_orient = gfapy.invert(tmp)
    self.overlap = self.overlap.complement()
    return self

File: lib/rgfa/line/edge/link/complement.rb
module RGFA::Line::Edge::Link::Complement

  # Creates the equivalent link with from/to inverted.
  #
  # The CIGAR operations (order/type) are inverted as well.
  # Tags are left unchanged.
  #
  # @note The path references are not copied to the complement link.
  #
  # @note This method shall be overridden if custom tags
  #   are defined, which have a ``complementation'' operation which determines
  #   their value in the equivalent complement link.
  #
  # @return [RGFA::Line::Edge::Link] the inverted link.
  def complement
    l = self.clone
    l.from = to
    l.from_orient = to_orient.invert
    l.to = from
    l.to_orient = from_orient.invert
    l.overlap = overlap.complement
    l
  end

  # Complements the link inplace.
  # The tags are left unchanged.
  #
  # @note The path references are not complemented by this method; therefore
  #   the method shall be used before the link is embedded in a graph.
  #
  # @note This method shall be overridden if custom tags
  #   are defined, which have a ``complementation'' operation which determines
  #   their value in the complement link.
  #
  # @return [RGFA::Line::Edge::Link] self
  def complement!
    tmp = self.from
    self.from = self.to
    self.to = tmp
    tmp = self.from_orient
    self.from_orient = self.to_orient.invert
    self.to_orient = tmp.invert
    self.overlap = self.overlap.complement
    return self
  end

end
